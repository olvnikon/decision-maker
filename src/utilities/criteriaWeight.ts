import { flow } from 'fp-ts/function';

type combinationsT = Array<Array<number>>;
type matrixT = Array<Array<number>>;
type geometricMeans = number[];

export const generateCombinations = (criteria: string[]) =>
  criteria.reduce((acc: combinationsT, _c, criteriaIndex) => {
    // -1 because the current index should be excluded
    const combinationLength = criteria.length - 1 - criteriaIndex;
    return combinationLength < 1
      ? acc
      : acc.concat(
          [...new Array(combinationLength)].map((_s, combinationIndex) => [
            criteriaIndex,
            // +1 because it should start from the next index after the current
            criteriaIndex + 1 + combinationIndex,
          ])
        );
  }, []);

const applyCriteriaValue = (combinationComp: number[]) => (combs: combinationsT): combinationsT =>
  combs.map((combination, index) => [...combination, combinationComp[index]]);

const generateMatrix = (combinationValues: combinationsT): matrixT => {
  const matrix = [...new Array(combinationValues.length)].map(() => [...new Array(combinationValues.length)]);
  // FIXME: matrix is generated incorrectly
  combinationValues.forEach(([i1, i2, value]) => {
    matrix[i1][i2] = value;
    matrix[i2][i1] = 1 / value;
  });
  matrix.forEach((_r, i) => {
    matrix[i][i] = 1;
  });
  return matrix;
};

const calculateGeometricMeans = (matrix: matrixT): geometricMeans =>
  matrix.map((row) =>
    Math.pow(
      row.reduce((acc, value) => acc * value, 1),
      1 / matrix.length
    )
  );

const calculateWeights = (gm: geometricMeans) =>
  gm.map((geometricMean) => geometricMean / gm.reduce((acc, gm) => acc + gm, 0));

export const criteriaToWeights = (combinationComp: number[]) =>
  flow(applyCriteriaValue(combinationComp), generateMatrix, calculateGeometricMeans, calculateWeights);
