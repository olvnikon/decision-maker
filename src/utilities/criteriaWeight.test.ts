import { generateCombinations } from './criteriaWeight';

const generateCriteria = (amount: number) => Array.from({ length: amount }, (_c, index) => `c${index}`);
const calculateAmountOfUniquePairs = (n: number) => (n * (n - 1)) / 2;

describe('generateCombinations', () => {
  test('should be [] if there is only 1 criteria', () => {
    expect(generateCombinations(generateCriteria(1))).toEqual([]);
  });

  test('should be only 1 combination when 2 criteria', () => {
    const criteria = generateCriteria(2);
    expect(generateCombinations(criteria)).toHaveLength(calculateAmountOfUniquePairs(2));
    expect(generateCombinations(criteria)).toEqual([[0, 1]]);
  });

  test('should be 3 combinations when 3 criteria', () => {
    const criteria = generateCriteria(3);
    expect(generateCombinations(criteria)).toHaveLength(calculateAmountOfUniquePairs(3));
    expect(generateCombinations(criteria)).toEqual([
      [0, 1],
      [0, 2],
      [1, 2],
    ]);
  });

  test('should match snapshot for many combinations', () => {
    const criteria = generateCriteria(7);
    expect(generateCombinations(criteria)).toMatchSnapshot();
    expect(generateCombinations(criteria)).toHaveLength(calculateAmountOfUniquePairs(7));
  });
});
