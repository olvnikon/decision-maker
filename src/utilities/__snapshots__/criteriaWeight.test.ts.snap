// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateCombinations should match snapshot for many combinations 1`] = `
Array [
  Array [
    0,
    1,
  ],
  Array [
    0,
    2,
  ],
  Array [
    0,
    3,
  ],
  Array [
    0,
    4,
  ],
  Array [
    0,
    5,
  ],
  Array [
    0,
    6,
  ],
  Array [
    1,
    2,
  ],
  Array [
    1,
    3,
  ],
  Array [
    1,
    4,
  ],
  Array [
    1,
    5,
  ],
  Array [
    1,
    6,
  ],
  Array [
    2,
    3,
  ],
  Array [
    2,
    4,
  ],
  Array [
    2,
    5,
  ],
  Array [
    2,
    6,
  ],
  Array [
    3,
    4,
  ],
  Array [
    3,
    5,
  ],
  Array [
    3,
    6,
  ],
  Array [
    4,
    5,
  ],
  Array [
    4,
    6,
  ],
  Array [
    5,
    6,
  ],
]
`;
